src/mdns/mdns_avahi.cpp

```diff
@@ -533,34 +629,35 @@ otbrError PublisherAvahi::PublishService(const char *   aHostName,
         used = static_cast<size_t>(reinterpret_cast<uint8_t *>(curr) - reinterpret_cast<uint8_t *>(buffer));
     }
 
-    for (Services::iterator it = mServices.begin(); it != mServices.end(); ++it)
+    serviceIt = FindService(aName, aType);
+
+    if (serviceIt == mServices.end())
     {
-        if (!strncmp(it->mName, aName, sizeof(it->mName)) && !strncmp(it->mType, aType, sizeof(it->mType)) &&
-            it->mPort == aPort)
-        {
-            otbrLog(OTBR_LOG_INFO, "MDNS update service %s", aName);
-            error = avahi_entry_group_update_service_txt_strlst(
-                mGroup, AVAHI_IF_UNSPEC, mProtocol, static_cast<AvahiPublishFlags>(0), aName, aType, mDomain, last);
-            SuccessOrExit(error);
-            ret = OTBR_ERROR_NONE;
-            ExitNow();
-        }
+        SuccessOrExit(ret = CreateService(*mClient, aName, aType, serviceIt));
+    }
+    else if (serviceIt->mHostName != safeHostName || serviceIt->mPort != aPort)
+    {
+        SuccessOrExit(ret = ResetGroup(serviceIt->mGroup));
+    }
+    else
+    {
+        otbrLog(OTBR_LOG_INFO, "MDNS update service %s for host '%s'", aName, safeHostName);
```

Comment:
nit:
```suggestion
        otbrLog(OTBR_LOG_INFO, "[mdns] update service %s for host '%s'", aName, safeHostName);
```

We use `[mdns]` in the mDNSResponder implementation, let's follow it :).